rfcs/0000-graphql-api.md

```diff
@@ -0,0 +1,433 @@
+  ownedWallets: [Wallet!]!
+  
+  wallet(publicKey: PublicKey!): Wallet
+  
+  payments(
+    filter: PaymentFilterInput,
+    first: Int,
+    after: String,
+    last: Int,
+    before: String): PaymentConnection
+  
+  blocks(
+    filter: BlockFilterInput,
+    first: Int,
+    after: String,
+    last: Int,
+    before: String): BlockConnection
+  
+  # Null if node isn't performing snark work
+  currentSnarkWorker: SnarkWorker
+
+  # Current sync status of the node
+  syncState: SyncUpdate!
+  
+  # version of the node (commit hash or version #)
+  version: String!
+  
+  # Network that the node is connected to
+  network: String
+  status: NodeStatus
+}
+
+type Mutation {
+  createPayment(input: CreatePaymentInput!): CreatePaymentPayload
+  
+  setSnarkWorker(input: SetSnarkWorkerInput!): SetSnarkWorkerPayload
+  
+  # Configure which network your node is connected to
+  setNetwork(input: SetNetworkInput!): SetNetworkPayload
+  
+  # Adds transaction to the node (note: Not sure how we want to represent this yet)
+  addPaymentReceipt(input: AddPaymentReceiptInput!): AddPaymentReceiptPayload
+  
+  # Tell server to track a private key and all associated transactions
+  addWallet(input: AddWalletInput!): AddWalletPayload
+  
+  # Deletes private key associated with `key` and all related information
+  deleteWallet(input: DeleteWalletInput!): DeleteWalletPayload
+}
+
+type Subscription {
+  # Subscribe to sync status of the node
+  newSyncUpdate: SyncUpdate!
+  
+  # Subscribe to payments for which this key is the sender or receiver
+  newPaymentUpdate(filterBySenderOrReceiver: PublicKey!): PaymentUpdate!
+  
+  # Subscribe all blocks created by `key`
+  newBlock(key: PublicKey): BlockUpdate!
+  
+  # Subscribe to fees earned by key
+  newSnarkFee(key: PublicKey): SnarkFee!
+}
+
+schema {
+  query: Query
+  mutation: Mutation
+  subscription: Subscription
+}
+```
+
+Staking is a little bit tricky because of the different states involved. These queries
+are pulled out for clarity, but you can imagine them being simply added to the above schema.
+
+You can either perform staking yourself, or delegate your stake to another public key
+- Changing your delegation status is a transaction,
+  which will experience the same consensus flow as other transactions,
+  with the addition of having to wait an additional epoch (?) for it to
+  actually come into effect.
+- Doing staking work yourself doesn't involve a transaction unless you need to
+  cancel an existing delegation first.
+
+This means that at any given moment, there could be a number of pending delegation
+transactions all awaiting consensus.
+
+``` graphql
+type Delegation {
+  nonce: UInt32!,
+  submittedAt: Date!
+  includedAt: Date
+  from: PublicKey!
+  to: PublicKey!
+  fee: UInt32!,
+  memo: String
+}
+
+type DelegationUpdate {
+  status: Delegation!
+
+  # We may have reached consensus but still be waiting for the correct epoch
```

Comment:
probably would need to subscribe to any changes that happens to the epoch ledger in order to implement this